substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,234 @@
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.Instrumentation;
+import java.lang.instrument.UnmodifiableClassException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarFile;
+
+/**
+ * Java agent can do instrumentation initialization work in premain phase. This class supports
+ * registering such premain methods at native image build time and invoking them at native image
+ * runtime. <br>
+ * JVM supports two kind of premain methods:
+ * <ol>
+ * <li>{@code public static void premain(String agentArgs, Instrumentation inst)}</li>
+ * <li>{@code public static void premain(String agentArgs)}</li>
+ * </ol>
+ * For the first one, at registration time we will set the second parameter with an instance of
+ * {@link SVMRuntimeInstrumentImpl} class which does not do any actual work as no instrumentation
+ * can do in native code at runtime.
+ * <p>
+ * <b>Be noticed</b>, the original agent premain method may not work well at native image runtime
+ * even if the input {@link Instrumentation} class is replaced with
+ * {@link SVMRuntimeInstrumentImpl}.
+ * </p>
+ * <p>
+ * It is the agent developers' responsibility to implement a native version of their agent premain
+ * method. It can be implemented in two ways:
+ * <ul>
+ * <li>Isolate code by checking current runtime. For example: <code>
+ *         <pre>
+ *         String vm = System.getProperty("java.vm.name");
+ *         if ("Substrate VM".equals(vm)) {
+ *           // native image runtime
+ *         } else{
+ *           // JVM runtime
+ *         }
+ *         </pre>
+ *     </code></li>
+ * <li>Use {@link com.oracle.svm.core.annotate.TargetClass} API to implement a native image version
+ * premain.</li>
+ * </ul>
+ * </p>
+ */
+public class PreMainSupport {
+
+    private static final String PREMAIN_OPTION_PREFIX = "-XX-premain:";
+
+    class PremainMethod {
+        String className;  // full qualified premain class name
+        Method method;
+        Object[] args;
+
+        PremainMethod(String className, Method method, Object[] args) {
+            this.className = className;
+            this.method = method;
+            this.args = args;
+        }
+    }
+
+    private Map<String, String> premainOptions = new HashMap<>();
+    // Order matters
+    private List<PremainMethod> premainMethods = new ArrayList<>();
+
+    @Platforms({Platform.HOSTED_ONLY.class})
+    public void registerPremainMethod(String className, Method executable, Object... args) {
+        premainMethods.add(new PremainMethod(className, executable, args));
+    }
+
+    /**
+     * Retrieve premain options from input args.
+     * Keep premain options and return the rest args as main args.
+     * The premain options format:
+     * <br>
+     * -XX-premain:[full.qualified.premain.class]:[premain options]
```

Comment:
I am not sure if this should be our final format for the flag. I think we can use it until we complete the feature and decide what is the user interface in the end.

Reply:
Agreed. I added a `//todo` for future changing.